dandi/cli/cmd_ls.py

```diff
@@ -99,39 +119,47 @@ def ls(paths, fields=None, format="auto", recursive=False):
     async_keys = tuple(async_keys.difference(common_fields))
 
-    process_paths = set()
+    process_assets = set()
     errors = defaultdict(list)  # problem: [] paths
     with out:
-        for path in files:
-            while len(process_paths) >= 10:
+        for asset in assets_gen():
+            while len(process_assets) >= 10:
                 lgr.log(2, "Sleep waiting for some paths to finish processing")
                 time.sleep(0.5)
-            process_paths.add(path)
 
-            rec = {}
-            rec["path"] = path
-
-            try:
-                if (not fields or "size" in fields) and not op.isdir(path):
-                    rec["size"] = os.stat(path).st_size
-
-                if async_keys:
-                    cb = get_metadata_pyout(
-                        path, async_keys, process_paths, flatten=format == "pyout"
-                    )
-                    if format == "pyout":
-                        rec[async_keys] = cb
-                    else:
-                        # TODO: parallel execution
-                        # For now just call callback and get all the fields
-                        for k, v in cb().items():
-                            rec[k] = v
-            except Exception as exc:
-                lgr.debug("Problem obtaining metadata for %s: %s", path, exc)
-                errors[str(type(exc).__name__)].append(path)
+            if isinstance(asset, str):  # path
+                process_assets.add(asset)
+                rec = {}
+                rec["path"] = asset
+
+                try:
+                    if (not fields or "size" in fields) and not op.isdir(asset):
+                        rec["size"] = os.stat(asset).st_size
+
+                    if async_keys:
+                        cb = get_metadata_pyout(
+                            asset, async_keys, process_assets, flatten=format == "pyout"
+                        )
+                        if format == "pyout":
+                            rec[async_keys] = cb
+                        else:
+                            # TODO: parallel execution
+                            # For now just call callback and get all the fields
+                            for k, v in cb().items():
+                                rec[k] = v
+                except Exception as exc:
+                    lgr.debug("Problem obtaining metadata for %s: %s", asset, exc)
+                    errors[str(type(exc).__name__)].append(asset)
+            elif isinstance(asset, dict):
+                # ready record
+                # TODO: harmonization for pyout
+                rec = asset
+            else:
+                raise TypeError(asset)
+
             if not rec:
-                errors["Empty record"].append(path)
-                lgr.debug("Skipping a record for %s since emtpy", path)
+                errors["Empty record"].append(asset)
+                lgr.debug("Skipping a record for %s since emtpy", asset)
```

Comment:
"Empty" is misspelled.

Reply:
Thank you! Away from laptop ATM, would you be so kind to push fix and merge if all tests are still good. I tuned up your other suggestion to become a proper suggestion for GitHub (see how) and merged - thanks!